CONFIGURATION.md
# Configuration
This guide contains all the information you need to configure `trusted-traveler-scheduler` to your specifications. An example file of the configuration can be found at [config.example.json](config.example.json).

## Table of Contents

- [Current Appointment Date](#current-appointment-date)
- [Locations](#locations)
- [Notifications](#notifications)
  - [Notification Level](#notification-level)
  - [Notification URLs](#notification-urls)
  - [Test Notifications](#test-notifications)
- [Retrieval Interval](#retrieval-interval)
- [Appointment Times](#appointment-times)
  - [Start Appointment Time](#start-appointment-time)
  - [End Appointment Time](#end-appointment-time)

## Current Appointment Date

Default: None

Type: String

This represents the date of your current appointment if you have one. If you do not have one, leave it as an empty string. This value is used to determine whether to notify you if a new appointment is found. If it is later than your current appointment date, it will not notify you.

```json
{
	"current_appointment_date": "January 1, 2024"
}
```

This above configuration will notify you if a new appointment is found for December 1, 2023, but will not notify you if an appointment is found for January 2, 2024.

**Note:** This must be in the format of Month Day, Year (e.g. January 1, 2024).

## Locations

Default: []

Type: Comma Seperated List or Integer

This represents the IDs of the enrollment centers you wish to monitor. This can either be a list of locations, or a singular location represented by an integer. This list is used in addition to whatever arguments you pass in at run-time of the script. For more information on locations, please see [LOCATIONS.md](LOCATIONS.md).

```json
{
  "location_ids": [ 5140,5444 ]
}
```

or

```json
{
    "location_ids": 5140
}
```

## Notifications

### Notification Level
Default: 1

Type: Integer

This indicates the notification sensitivity you wish to receive. 
```json
{
  "notification_level": 1
}
```
Level 1 means you receive notifications when new appointments are found.

Level 2 means you receive notifications only when errors occur.

### Notification URLs

Default: []

Type: List or String

This uses the [Apprise Library][0] to generate notifications that are pushed to you based on the notification level you have elected to receive. You can find more information about notifications through the [Apprise README.md][1]

```json
{
  "notification_urls": "service://my_service_url"
}
```

or

```json
{
  "notification_urls": [
    "service://my_first_service_url",
    "service://my_second_service_url"
  ]
}
```

### Test Notifications
To test your notification configuration, run the following command:
```shell
python ttp.py -t
```

## Retrieval Interval
Default: 5 minutes

Type: String

This indicates how often the script will fetch new appointments from the monitored locations. To disable automatic retrieval, set this to "0m".

```json
{
    "retrieval_interval": "5m"
}
```

## Appointment Times

### Start Appointment Time
Default: 00:00

Type: String

This indicates the earliest appointment you would like to be notified for. To be notified for all appointments, set this to "00:00".

```json
{
    "start_appointment_time": "06:00"
}
```

### End Appointment Time
Default: 23:59

Type: String

This indicates the latest appointment you would like to be notified for. To be notified for all appointments, set this to "23:59".

```json
{
    "end_appointment_time": "20:00"
}
```



[0]: https://github.com/caronc/apprise
[1]: https://github.com/caronc/apprise#supported-notifications


src/config.py
import json
import re
import os
import sys
from typing import Any, Dict
from datetime import datetime

from .notification_level import NotificationLevel

CONFIG_FILE_NAME = "config.json"
LOCATION_FILE_NAME = "locations.json"

class Config:
    """
    A class representing the configuration for the Trusted Traveler Scheduler application.
    """
    def __init__(self):
        # Default values are set
        self.current_appointment_date = None
        self.location_ids = []
        self.notification_level = NotificationLevel.INFO
        self.notification_urls = []
        self.retrieval_interval = 300
        self.start_appointment_time = datetime(year=9999, month=12, day=31, hour=0, minute=0)
        self.end_appointment_time = datetime(year=9999, month=12, day=31, hour=23, minute=59)
       
        # Read the config file
        config = self._get_config()
        self.locations = self._get_locations()

        # Set the configuration values if provided
        try:
            self._parse_config(config)
        except TypeError as err:
            print("Error in configuration file:")
            print(err)
            sys.exit()

    def _get_config(self) -> Dict[str, Any]:
        """
        Reads the configuration file and returns its contents as a dictionary.

        Returns:
            A dictionary containing the configuration values.
        """
        project_dir = os.path.dirname(os.path.dirname(__file__))
        config_file = project_dir + "/" + CONFIG_FILE_NAME

        config = {}
        try:
            with open(config_file) as file:
                config = json.load(file)
        except FileNotFoundError:
            pass

        return config
    
    def _get_locations(self) -> Dict[str, Any]:
        """
        Reads the locations file and returns its contents as a dictionary.

        Returns:
            A dictionary containing the locations.
        """
        project_dir = os.path.dirname(os.path.dirname(__file__))
        locations_file = project_dir + "/utils/" + LOCATION_FILE_NAME

        locations = {}
        try:
            with open(locations_file, encoding="utf-8") as file:
                locations = json.load(file)
        except FileNotFoundError:
            pass

        return locations
    
    # This method ensures the configuration values are correct and the right types.
    # Defaults are already set in the constructor to ensure a value is never null.
    def _parse_config(self, config: Dict[str, Any]) -> None:
        """
        Parses the configuration dictionary and sets the corresponding attributes of the Config object.

        Args:
            config: A dictionary containing the configuration values.

        Raises:
            TypeError: If any of the configuration values are of the wrong type.
            ValueError: If any of the configuration values are invalid.
        """
        if "current_appointment_date" in config and config["current_appointment_date"] != "":
            self.current_appointment_date = config["current_appointment_date"]

            try:
                self.current_appointment_date = datetime.strptime(self.current_appointment_date, '%B %d, %Y')
            except:
                raise TypeError("'current_appointment_date' must be in the format of Month Day, Year (e.g. January 1, 2024)")
        
        if "location_ids" in config:
            self.location_ids = config["location_ids"]

            if not isinstance(self.location_ids, (list, int)):
                raise TypeError("'location_ids' must be a list or integer")
            
        if "notification_level" in config:
            self.notification_level = config["notification_level"]

            if not isinstance(self.notification_level, int):
                raise TypeError("'notification_level' must be an integer")

        if "notification_urls" in config:
            self.notification_urls = config["notification_urls"]

            if not isinstance(self.notification_urls, (list, str)):
                raise TypeError("'notification_urls' must be a list or string")

        if "retrieval_interval" in config:
            self.retrieval_interval = config["retrieval_interval"]

            if not isinstance(self.retrieval_interval, str):
                raise TypeError("'retrieval_interval' must be a string")
            
            try:
                self.retrieval_interval = self.convert_to_seconds(self.retrieval_interval)
            except ValueError as err:
                raise TypeError(err)
        
        if "start_appointment_time" in config:
            self.start_appointment_time = config["start_appointment_time"]

            if not isinstance(self.start_appointment_time, str):
                raise TypeError("'start_appointment_time' must be a string")
            
            try:
                self.start_appointment_time = self.convert_to_datetime(self.start_appointment_time)
            except ValueError as err:
                raise TypeError(err)
            
        if "end_appointment_time" in config:
            self.end_appointment_time = config["end_appointment_time"]

            if not isinstance(self.end_appointment_time, str):
                raise TypeError("'end_appointment_time' must be a string")
            
            try:
                self.end_appointment_time = self.convert_to_datetime(self.end_appointment_time)
            except ValueError as err:
                raise TypeError(err)
    
    def convert_to_seconds(self, time: str) -> int:
        """
        Converts a time string to seconds.

        Args:
            time: A string representing a time interval in the format of <integer><unit>. (e.g. 45s (seconds), 30m (minutes), 2h (hours), 1d (days))

        Returns:
            An integer representing the time interval in seconds.

        Raises:
            ValueError: If the time string is not in the correct format or contains an invalid time unit.
        """
        # If the time is already an integer, return it
        try:
            return int(time)
        except:
            pass

        match = re.match(r'^(\d+)([smhd])$', time.lower())
        
        if not match:
            raise ValueError(f"'retrieval_interval' must be in the format of <integer><unit>. (e.g. 45s (seconds), 30m (minutes), 2h (hours), 1d (days))")
        
        value, unit = int(match.group(1)), match.group(2)

        if unit == "s":
            return value
        elif unit == "m":
            return value * 60
        elif unit == "h":
            return value * 3600
        elif unit == "d":
            return value * 86400
        else:
            raise ValueError(f"'retrieval_interval' invalid time unit: {unit}. Accepted units: s (seconds), m (minutes), h (hours), d (days).")
        
    def convert_to_datetime(self, time: str) -> datetime:
        """
        Converts a time string to a datetime object.

        Args:
            time: A string representing a time in the format of HH:MM.

        Returns:
            A datetime object representing the time.

        Raises:
            ValueError: If the time string is not in the correct format.
        """
        return datetime.strptime(time, "%H:%M")
        
        
